tests/python_package_test/test_engine.py

```diff
@@ -2278,6 +2311,99 @@ def test_multiple_feval_train():
     assert 'decreasing_metric' in evals_result['valid_0']
 
 
+def test_objective_callable_train():
```

Comment:
can you please split these new tests up by learning task?

* `test_objective_callable_train_regression`
* `test_objective_callable_train_binary_classification`
* `test_objective_callable_cv_regression`
* `test_objective_callable_cv_binary_classification`

Since I don't see any logic shared between the regression and classification code proposed in these new tests, I think splitting them up is preferable. Finer-grained tests make debugging a bit faster, as "all of the custom-objective-function tests are failing" has different possible root causes than, say, "only the regression custom-objective-function tests are failing".

That pattern has been really helpful for me in working on `lightgbm.dask`. For example

https://github.com/microsoft/LightGBM/blob/17d4e007852b4f0b34b211a674cb26e6751c98e2/tests/python_package_test/test_dask.py#L443

means that I've been able to run _just that test_ like

```shell
pytest tests/python_package_tests/test_dask.py::test_classifier_custom_objective
```

which has really sped up the testing cycle for some types of changes

Reply:
makes sense that tests check a single test case, thanks for providing references, it helps with the learning.

Reply:
Great! I just marked this thread "unresolved". Please don't mark cobversations resolved before you've puhed the related changes, as that mkes it difficult for reviewers to understand the relationship between comments and commits, and to confirm that all suggestions have been addressed.

Reply:
got it